### ISSUE TITLE
Invalid SAML response

### ISSUE THREAD

**Issue by bruno222** [tngan/samlify#466]
Any clue on why this SAML Response is invalid? Every 10 logins, I can login around 4 times... The rest gives me bad SAML response... This is a custom template for SAML Response.

And if I go to [any SAML tool](https://www.samltool.com/validate_response.php), it actually says me the SAML is invalid.. But I compared both answers (a good one and this one below) using Git Diff and I can only see trivial things being different (ID, InResponseTo, Signature)... But the structure is exact the same!

What mistake I am doing?

```
<?xml version="1.0" encoding="UTF-8"?>
<saml2p:Response 
    Destination="https://iam.redacted.com/v1/Accounts/redacted/saml2" 
    ID="00ed9a13-a7b5-4a91-9da3-e22455df2c73" 
    InResponseTo="ONELOGIN_7c2a509b-b02f-446a-8dab-738e8f838334" 
    IssueInstant="2022-01-31T15:36:37.967Z" 
    Version="2.0" 
    xmlns:saml2p="urn:oasis:names:tc:SAML:2.0:protocol" 
    xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <saml2:Issuer 
        Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity" 
        xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">https://redacted.eu.ngrok.io/sso/saml
    </saml2:Issuer>
    <ds:Signature 
        xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
        <ds:SignedInfo>
            <ds:CanonicalizationMethod 
                Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
            <ds:SignatureMethod 
                Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
            <ds:Reference 
                URI="#00ed9a13-a7b5-4a91-9da3-e22455df2c73">
                <ds:Transforms>
                    <ds:Transform 
                        Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
                    <ds:Transform 
                        Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
                </ds:Transforms>
                <ds:DigestMethod 
                    Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
                <ds:DigestValue>
                    ki6vlOtredactedlp5ACmjoTJaR0=
                </ds:DigestValue>
            </ds:Reference>
        </ds:SignedInfo>
        <ds:SignatureValue>redacted+redacted+w/Tc1rR7FMHPWJlVNL8L9pr0hiDkkBadxGAYdVYD15jtbCa44Mn899AatZ9oLEQyd7g92aI6VjIrlV4yIB7DP+YnpHyI5apCM/
            zc=
        </ds:SignatureValue>
        <ds:KeyInfo>
            <ds:X509Data>
                <ds:X509Certificate>MIICfTCCAeYCCQD+JB1LihW1ZTANBgkqhkiG9w0BAQsFADCBgjELMAkGA1UEBhMC
REUxDzANBgNVBAgMBk11bmljaDEPMA0GA1UEBwwGTXVuaWNoMQ8wDQYDVQQKDAZU
d2lsaW8xDzANBgNVBAsMBlR3aWxpbzEPMA0GA1UEAwwGVHdpbGlvMR4wHAYJKoZI
hvcNAQkBFg9ibGFoQHR3aWxpby5jb20wHhcNMjIwMTMxMDAzMjM3WhcNMjMwMTMx
MDAzMjM3WjCBgjELMAkGA1UEBhMCREUxDzANBgNVBAgMBk11bmljaDEPMA0GA1UE
redacted
redacted
redacted/5VgHFCqujDNEIonf
M18tFTAKvKHvb56qp9BJORimTDfOpTvxuE7hOrEpvZqmayfdwYYTzy7JxdmWzWtU
/oS6PsvcPLLagWPP1u2/Z6IGgedRCpUAbAfXeNLf1Pbc/WGezF3MyiZ4F86ORJlD
W4nnGU7RVOGVAgMBAAEwDQYJKoZIhvcNAQELBQADgYEADBAarm5v7UUP7BlheQ+2
FmTaqf91n81gHrz3nmIkiXNZ+ZW90RlFqUTka1gREdZj9uk+oEz/yU4v3jXho2Pl
twpZtVwcEUrvktKewFuusgErwRetu+Qv9ysQu4SY5Q7j2ayAAvFETbTnyWcEk34l

                    CMKOuKHSpD7wVRhUxdcthCY=
                </ds:X509Certificate>
            </ds:X509Data>
        </ds:KeyInfo>
    </ds:Signature>
    <saml2p:Status 
        xmlns:saml2p="urn:oasis:names:tc:SAML:2.0:protocol">
        <saml2p:StatusCode 
            Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
    </saml2p:Status>
    <saml2:Assertion 
        ID="_f415c00b-18b5-4157-9b2f-b2dca4976a79" 
        IssueInstant="2022-01-31T15:36:37.967Z" 
        Version="2.0" 
        xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion" 
        xmlns:xs="http://www.w3.org/2001/XMLSchema">
        <saml2:Issuer 
            Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity" 
            xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">https://redacted.eu.ngrok.io/sso/saml
        </saml2:Issuer>
        <saml2:Subject 
            xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">
            <saml2:NameID 
                Format="urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified">bkilian+agent1@redacted.com
            </saml2:NameID>
            <saml2:SubjectConfirmation 
                Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
                <saml2:SubjectConfirmationData 
                    InResponseTo="ONELOGIN_7c2a509b-b02f-446a-8dab-738e8f838334" 
                    NotOnOrAfter="2022-01-31T15:44:37.967Z" 
                    Recipient="https://iam.redacted.com/v1/Accounts/redacted/saml2"/>
            </saml2:SubjectConfirmation>
        </saml2:Subject>
        <saml2:Conditions 
            NotBefore="2022-01-31T15:36:37.967Z" 
            NotOnOrAfter="2022-01-31T15:44:37.967Z" 
            xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">
            <saml2:AudienceRestriction>
                <saml2:Audience>https://iam.redacted.com/v1/Accounts/redacted/saml2/metadata</saml2:Audience>
            </saml2:AudienceRestriction>
        </saml2:Conditions>
        <saml2:AuthnStatement 
            AuthnInstant="2022-01-31T15:36:37.967Z" 
            SessionIndex="ONELOGIN_7c2a509b-b02f-446a-8dab-738e8f838334" 
            xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">
            <saml2:AuthnContext>
                <saml2:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport</saml2:AuthnContextClassRef>
            </saml2:AuthnContext>
        </saml2:AuthnStatement>
        <saml2:AttributeStatement 
            xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">
            <saml2:Attribute 
                Name="full_name" 
                NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
                <saml2:AttributeValue 
                    xmlns:xs="http://www.w3.org/2001/XMLSchema" 
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
                    xsi:type="xs:string">Agent1 Hello2
                </saml2:AttributeValue>
            </saml2:Attribute>
            <saml2:Attribute 
                Name="email" 
                NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
                <saml2:AttributeValue 
                    xmlns:xs="http://www.w3.org/2001/XMLSchema" 
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
                    xsi:type="xs:string">bkilian+agent1@redacted.com
                </saml2:AttributeValue>
            </saml2:Attribute>
            <saml2:Attribute 
                Name="roles" 
                NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
                <saml2:AttributeValue 
                    xmlns:xs="http://www.w3.org/2001/XMLSchema" 
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
                    xsi:type="xs:string">supervisor, wfo.full_access
                </saml2:AttributeValue>
            </saml2:Attribute>
        </saml2:AttributeStatement>
    </saml2:Assertion>
</saml2p:Response>
```

**Comment by carsonwah** [comment]
It is possibly due to time out-of-sync issue. It fails randomly maybe because the network latency differs every time.

You may try adding `clockDrifts` option to tolerate a time range. See if it helps.
https://samlify.js.org/#/sp-configuration?id=optional-parameters

**Comment by bruno222** [comment]
Thank you, @carsonwah! I have found the issue!

- Before: `ID="00ed9a13-a7b5-4a91-9da3-e22455df2c73"`

- After: ``ID="blah_00ed9a13a7b54a919da3e22455df2c73"``

I read somewhere that the ID has to follow this pattern and once I changed, bammm, started to work 100% of the time! So it seems the **ID** attribute has to be `anything_anything` - with this mandatory underline.... Strange, but OK, it now works 100% of the time! Thanks once again @carsonwah.
